Checkpoint initial version

fbeltrao opened the pull request:
The goal of this PR is to start the conversion about the checkpoint implementation. Ideally we work together until a first version is accepted. We build on top of it with changes in other PRs.

**Batch item functions**
- Each partition is handled individually
- Incoming messages are added to queue. Once queue length is equal to batch size it is written to channel
- Channel reader calls the function, and commits the current offset
- If the channel is full, the partition consumption is paused. It is resumed if by the channel reader

**Single item functions**
- All partitions are handled together
- Incoming messages are added to queue. Once queue length is equal to batch size it is written to channel
- Channel reader calls the function, and commits all partition/offset combinations in processed batch
- If the channel is full, the partition consumption is paused. It is resumed if by the channel reader

**Not included**
- Handling functions failures
- Handling topic with few items (this should trigger the function call based on time as opposed to queue having enough items).

ryancrawcour commented:
_"Incoming messages are added to queue. Once queue length is equal to batch size it is written to channel"_

What happens if hours go by before the queue length equals batch size? 
Do we get no events in to the channel, and therefore no Function triggered, until this happens?

ryancrawcour commented:
Never mind, I see you answered my questions on "circuit breaker" patter in the **"Not Included"** note. We should add an issue to track this separately then and ensure it is added.

jeffhollan commented:
I'm slightly confused by a few pieces in this PR.  I think the term "publisher" is throwing me off.  At first I assumed that was the output binding (e.g. 'publishing' messages to push to Kafka), but that doesn't seem to be the case? Can you clarify what a publisher here does?  The other term I'm slightly confused by is why we're using "channels" directly here instead of the concept I'm a bit more familiar with of "subscriptions" / "consumers."  

Aside from that:
1. It almost reads like the code is reading or dealing with everything message by message, but creating it's own in-memory batches?  That seems inefficient.  I would hope the case for a consumer is it pulls a batch directly from Kafka (1 call to Kafka results in 50 events or something) and then deals with it directly.  I'm not sure why we'd need to fill our own batch or determine the batch size.  Also I expect what we actually define is a max batch size. It's very possible that the batch may be much less than the max, but that's ok. Our goal is to process as quickly as possible rather than wait for and hold messages.

1. Similarly, the concept of max batches.  This works, but in event hubs it's configurable and called [the batchCheckpointFrequency](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-hubs#host-json), meaning how many batches should we process before making the checkpoint.

1. In general the code feels much more involved than I expected it would be.  Most of this is due to the channel / publisher confusion I had, but expected it would have looked a lot more like [this first example in the Confluent SDK](https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/examples/Consumer/Program.cs) of creating a simple subscription consumer, and then checkpointing.  Instead of `consumer.Consume()` I was hoping there would be a `consumer.ConsumeBatch()`.  Maybe Kafka doesn't have that concept built in like Event Hubs which is fine, hopefully it doesn't hurt throughput, but I'd also still expect something closer to that example.

fbeltrao commented:
@jeffhollan , thanks a lot for your comments.

You are right about naming confusion. Naming the publisher as "function executor" or "dispatcher" makes more sense.

Processing a message happens the following way:

1. A Kafka listener connects a consumer (class from librdkafka) with a function binding. It creates the ~~publisher~~ function executor which calls the function
1. Listener calls "Consume", receiving 1 message at a time (!!), sending to the ~~publisher~~ function executor
1. The function executor, as you observed, keeps messages in memory. Once a full batch is ready it is written to a channel
1. Channel reader gets the packets that are ready for consumption and calls the function. After completing all calls, it commits the offset and waits for the next batch.

**Why are we consuming 1 event at a time?**
As far as I have seen, this is how the functionality in exposed in the [library](https://github.com/confluentinc/confluent-kafka-dotnet).
Even in their [C++ sample](https://github.com/edenhill/librdkafka/blob/master/examples/rdkafka_consume_batch.cpp#L97) a "batch consumer" is just calling "consume" multiple times. Maybe there is a way, I just did not find it. I have the same concerns as you, picking one message at a time does not look good. As of now a function host would have a single consumer for a given topic. We could experiment having multiple consumers,  one per partition in order to increase the amount of parallel "consume" calls we do. However, I am not sure it provides a better throughput (would have to dig deeper in librdkafka), and it would mean we need to add our own partition leasing logic, which is given by the library. Another approach is to mimic their example and have the multiple "consume" calls for a certain amount of time and process them as a batch (like the C++ example).

**Why did I use a channel?**
Two reasons:
1. I did not want to have the function execution blocking the "getting messages from Kafka".
1. Channels provides a simple way to ensure I execute functions in partition order with a minimal amount of code.

**My proposal**
In order to minimise the local state and handle messages as fast possible, but still using the library contract we have, I change the subscriber implementation to mimic what their sample code does. Essentially calling "consume" in a loop until a maxBatchSize or a timeout is reached. This change would make the code simpler and eliminate the need to keep a long lists of messages.

I then see two options for the subscriber and function executor. Should they be:
a) Synchronous: get batch &rarr; execute function &rarr; commit &rrar; repeat
b) Asynchronous: get batch &rarr; write to channel &rarr; repeat. While function executor is waiting for items in the channel.

Option b seems better to me.

What do you think?

jeffhollan commented:
Awesome explanation.  I think I'm ok with either a or b proposed assuming that:
1. Users can write a function in a way that makes sure they will process messages in order. In other words I don't have to worry about a concurrent execution on my instance of message "B" when I just got message "A"
1. We do what we can to work around the batch semantics (or lack thereof) to allow for good perf

That helps a lot, so with that I think I'm cool with the options as you laid them out.

fbeltrao commented:
So it means that for a single item trigger function I should not make a Task.WhenAll for all messages in same partition, right?

I must, in a loop do
  Execute  func item[i]
  Commit item[i]

But different partitions can run in parallel.

jeffhollan commented:
Correct - though I expect the right thing is we have some counter with a default of like 30 and we commit every 30 message executions.  This means it's possible on execution 28 the instance halts and you'd have to replay the 28 you'd processed, but that's generally expected failure behavior with things like event hubs.  

We could execute different partitions on the same instance (assuming we have locks for multiples), but per [our other triggers](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) we generally like people to control the concurrency limit per instance.  So even if I had 100 partitions and a single instance I would still have some control to say something like "don't let more than 50 concurrent executions at any time on this instance" if that makes sense

ryancrawcour commented:
i am really surprised you cannot read a batch of messages from a Kafka topic, at once. that seems truly bizarre to me.

ryancrawcour commented:
I thought we are going to be _"having multiple consumers on a topic, one per partition"_ was this not one of your requirements @jeffhollan ?

To me it sounds like a combo of issues #4  and #5 ?

jeffhollan commented:
Yeah that's probably my bad in wording. The requirement would be at *most* one instance per partition.  Meaning we shouldn't (and I don't even think Kafka will let us) have 2 instances pulling from the same partition.  But very likely at some point multiple partitions will point to a single instance.  For low volume scenarios 1 instance may be enough, and it could pull from all partitions.

ryancrawcour commented:
So just to be sure I understand this then …

1 Topic can have multiple partitions

1 Function instance will lock to a single partition (meaning no other instance can touch it, until it releases, or crashes). 

1 Function instance can process multiple partitions (from the same topic)

1 Function instance can only process a single Topic, if 2 topics are configured we will have, at least, 2 Function instances running at any point in time. 

Is this correct?

ryancrawcour commented:
On the "Can you read a batch of messages from a Kafka topic at once" I found this SO post that says you can.

fbeltrao commented:
Can Confluent.Kafka read multiples? Do you have the link?

jeffhollan commented:
I think it's more:

- A Kafka topic can have multiple partitions
- A function instance can lock one or many partitions at one time
- A function instance can process from one or many partitions at one time (as long as other instances haven't locked that partition for themselves)
- It's possible you could have multiple functions inside the same app, each triggering from a different topic.  So in theory the same instance COULD be triggering from multiple topics.  *HOWEVER* a single function trigger could only trigger off of a single topic.

ryancrawcour commented:
@fbeltrao this [SO ](https://stackoverflow.com/questions/45920608/how-to-read-batch-messages-in-confluent-kafka-python) post seems to say the same thing you are though. It's the underlying library that confluent uses that reads 1 by 1.

ryancrawcour commented:
So, do we think this PR is ready to merge as is, or does it need some revision?

fbeltrao commented:
No, I am working on a version that is similar to what @jeffhollan and me agreed.

ryancrawcour commented:
Ok, great. Will leave the PR open and wait for updates then. Thanks @fbeltrao

fbeltrao commented:
I have updated the implementation. In a short:

Triggering function calls starts at the KafkaListener class. It is responsible for connecting a Kafka trigger function with a Kafka Consumer (from Confluent.Kafka library).
As the library does not seem to support a way to receive message in batches the current implementation loops getting items with a timeout, triggering the functions as messages are received.
At the listener startup a "Kafka subscriber thread" is created. 

The subscriber thread does the following:

```
while (!cancellationToken.IsCancellationRequested)
    while !MaxBatchReleaseTime is passed
        kafkaItem = read_from_kafka()
        pending_items[partition].add(kafkaItem)
        
        if (pending_items[partition].length > maxBatchSize)
            pending_items[partition].flush() // making the items available for the function executor
    end
    
    foreach partition not already flushed
        pending_items[i].flush()
end
```

Executing the function:

The function is executed by two type of executors: SingleItem and MultiItem.
A single item executor calls the received items in a loop, one by one. The multi item executor calls the function a single time.
Once all items have been processed the checkpoint is saved.

Items processing is implemented through a channel. Once the channel is full (by default 10 batches) the subscriber will pause until the function catches up. This will cause all partitions to wait.

ryancrawcour commented on docs/DevGuide.md:
```diff
@@ -16,7 +16,7 @@ A sample function is provided in folder sample/KafkaFunctionSample. It depends o
   "IsEncrypted": false,
   "Values": {
-    "AzureWebJobsStorage": "<storage-account>",
+    "AzureWebJobsStorage": "None",
```
:

ryancrawcour commented on docs/DevGuide.md:
```diff
@@ -62,3 +62,35 @@ Short description of building a web jobs trigger:
+        read from kafka
+        pending_items[partition].add(kafkaItem)
+        
+        if (pending_items[partition].length > maxBatchSize)
```
should this be >= ?

jeffhollan commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/MultipleItemFunctionExecutor.cs:
```diff
@@ -0,0 +1,73 @@
+
+
+                        var functionResult = await this.ExecuteFunctionAsync(triggerData, cancellationToken);
+                        if (functionResult.Succeeded)
```
I don’t think we want this check. We want to commit even if execution failed (for now at least - in future may add some retry logic). Event Hubs today also doesn’t make this check

https://github.com/Azure/azure-webjobs-sdk/blob/6930c4ea9ec78db12a05a2d0bf38cce42ab04e13/src/Microsoft.Azure.WebJobs.Extensions.EventHubs/Listeners/EventHubListener.cs#L176

fbeltrao replied:
Ok, it makes sense. I will make the required changes.

jeffhollan commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/KafkaListener.cs:
```diff
@@ -65,64 +80,35 @@ public void Cancel()
                     if (e.IsAssignment)
                     {
+                        foreach (var tp in e.Partitions)
+                        {
```
Is there any way we can control how many of these listeners spin up? My concern is that if we spin up a listener per partition and there are like 100 partitions. If I had a batch size of like 30 and all 100 partitions were active my function may try to execute 300 messages at a time and overwhelm itself

fbeltrao replied:
I am not sure I understand. There is a single KafkaListener for each function trigger. It will subscribe to one topic, owning between 0 and N partitions (depending how many other function hosts are active). The "read message loop" gets message from all owned partitions.

Would you like to limit the amount of topics a single function host can own?

jeffhollan replied:
No what you described is right. I wasn’t sure if this would spin up a background thread per partition but doesn’t sound like it.

fbeltrao commented:
@jeffhollan 

Just to make sure I understand your concept of triggering the functions. Using your data example:
Partition 1: ABCDE
Partition 2: 12345

They will be received by subscriber in following order: 
00:00:01: AB12C
00:00:02: 34DE5

**For batch items I would:**
1. Send to function batch containing AB12C
2. Commit C and 2
--
3. Send to function batch containing 34DE5
4. Commit D and 5

I MUST ensure that step 3 will only be executed after 2.


**For single items I would:**
1. Send A 
1. Send 1
2. Send B
2. Send and commit 2
3. Send and commit C
--
4. Send 3
4. Send D
5. Send 4
5. Send and commit E
6. Send and commit 5

jeffhollan commented:
Correct. At least that mirrors what happens in Event Hubs trigger today

fbeltrao commented:
Alright, thanks. I will adapt the code later today.

jeffhollan commented:
In fact just to be specific single it would be 

Send A
Send 1
Send 2
Send B
Send C
Commit C and 2 (must make sure this happens after all above have completed)

fbeltrao commented:
I have pushed the following changes

- No more function executor per partition since batches can contain items from distinct partitions
- Fixed gracefull shutdown
- Implemented order as described below. For single item functions I am guaranteeing that items in a partition are executed only once the previous item call finished.
I still run items from different partitions in parallel, but for a given partition it will be sequential. @jeffhollan, in your comments you said that eventhubs run all in parallel. This behavior is simpler to implement actual. Up to you: guaranteed order or run all in parallel?

### Execution order

**Data in Kafka**

```
Partition 1: ABCDE
Partition 2: 12345
```

**Receiving order**
```
00:00:01: AB12C
00:00:02: 34DE5
```

**Multi item function**

1. Batch AB12C is received
1. Send to function batch containing AB12C
1. Commit C and 2
1. Batch 34DE5 is received
1. Send to function batch containing 34DE5
1. Commit E and 5

**Single item function**

1. Batch AB12C is received
1. Sends A, 1 (in parallel)
1. Sends B, 2 (in parallel)
1. Sends C
1. Commits C and 2
1. Batch 34DE5 is received
1. Sends D, 3 (in parallel)
1. Sends E, 4 (in parallel)
1. Sends 5
1. Commits E and 5

jeffhollan commented:
This is fine. In event hubs today we only guarantee order on the batch case and for single message we don't - and I believe choice was to try to optimize throughput of single events.  But in my mind it's fine either way, and if people aren't happy with the throughput of single execution in the way you've created here they can always move to batch.  So I'm good with either one - this sounds good

fbeltrao commented:
@jeffhollan thanks for your patience and support!

fbeltrao commented:
Fixed a bug where the channel reader OperationCanceledException was not being handled during a host shutdown

fbeltrao commented on docs/DevGuide.md:
```diff
@@ -62,3 +62,35 @@ Short description of building a web jobs trigger:
+        read from kafka
+        pending_items[partition].add(kafkaItem)
+        
+        if (pending_items[partition].length > maxBatchSize)
```
It should never be greater, but I changed just in case

fbeltrao commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/FunctionExecutorBase.cs:
```diff
@@ -0,0 +1,126 @@
+    public abstract class FunctionExecutorBase<TKey, TValue> : IDisposable
+    {
+        const int ChannelWriteRetryInterval = 50;
+        const int ChannelCapacity = 10;
```
They will be configurable, next PR will address it.

fbeltrao commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/FunctionExecutorBase.cs:
```diff
@@ -0,0 +1,126 @@
+
+        protected Task<FunctionResult> ExecuteFunctionAsync(TriggeredFunctionData triggerData, CancellationToken cancellationToken)
+        {
+            // TODO: add retry logic
```
I think this is part of #17 , I will add comments there.

fbeltrao commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/KafkaListener.cs:
```diff
@@ -74,15 +90,24 @@ public Task StartAsync(CancellationToken cancellationToken)
 
                         this.logger.LogInformation($"Assigned partitions: [{string.Join(", ", e.Partitions)}]");
                         // possibly override the default partition assignment behavior:
                         // consumer.Assign(...) 
```
No need to revisit, will remove the comments.

fbeltrao commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/KafkaListener.cs:
```diff
@@ -174,28 +207,61 @@ private ConsumerConfig GetConsumerConfiguration()
+                    alreadyFlushedPartitions.Clear();
                 }
             }
             catch (OperationCanceledException)
```
Good catch, that is from the original version where the exception could be raised. I changed for a simple error logging catch.
